driver/builder_makefile.ml

```diff
@@ -192,12 +204,48 @@ let mk_gen_prepare gen =
   let rmk = `Blank :: `Comment header :: `Blank :: gen.rmk in
   Path.Map.fold add_prepare gen.preps { gen with rmk }
 
+let mk_all gen =
+  let add gen p = match Part.kind p with
+  | `Lib -> { gen with all = str "lib-%s" (Part.name p) :: gen.all }
+  | `Bin -> { gen with all = str "bin-%s" (Part.name p) :: gen.all }
+  | _ -> gen
+  in
+  let all gen = match gen.all with
+  | []  -> gen
+  | all ->
+      let header = "Main rule" in
+      let rule = Makefile.rule ~targets:["all"] ~prereqs:all ~recipe:[] () in
+      let rmk = `Blank :: rule :: `Comment header  :: gen.rmk in
+      { gen with rmk }
+  in
+  List.fold_left add gen (Project.parts gen.proj)
+  |> all
+
+let mk_test gen =
+  let add gen p = match Part.kind p with
+  | `Run -> { gen with test = str "run-%s" (Part.name p) :: gen.test }
+  | _ -> gen
+  in
+  let test gen = match gen.test with
+  | []    -> gen
+  | tests ->
+      let header = "Run tests" in
+      let rule = Makefile.rule ~targets:["test"] ~prereqs:tests ~recipe:[] () in
+      let rmk = `Blank :: rule :: `Comment header :: gen.rmk in
+      { gen with rmk }
+  in
+  List.fold_left add gen (Project.parts gen.proj)
+  |> test
+
 let of_project ~setup_files proj =
-  let gen = generator proj in
-  let gen = keys gen in
-  let gen = List.fold_left mk_part gen (Project.parts gen.proj) in
-  let gen = mk_gen_dirs gen in
-  let gen = mk_gen_ds gen in
-  let gen = mk_gen_phony gen in
-  let gen = mk_gen_prepare gen in
-  List.rev gen.rmk
+  let ret gen = List.rev gen.rmk in
+  generator proj
+  |> keys
+  |> mk_phony
+  |> mk_all
+  |> mk_test
+  |> (fun gen -> List.fold_left mk_part gen (Project.parts gen.proj))
+  |> mk_dirs
+  |> mk_ds
+  |> mk_prepares
+  |> ret
```

Comment:
Personally I don't find that kind of code more readable than the previous style. Point-free style has sometimes its uses in DSLs but in general it puts a larger cognitive burden on me (and has the effect of introducing needless functions in that case, and why name `ret` and not the other ?).  I prefer to read OCaml functions as sequences of `let` bindings where variables are explicitely threaded.